File: examples/python_tools/simple_calculator.py
#!/usr/bin/env python3
"""
Simple calculator tool for the workflow toolkit.
This script demonstrates how to create a Python tool that can be used with the plugin system.
"""

import json
import sys
from typing import Dict, Any, Union


def add(a: Union[int, float], b: Union[int, float]) -> Union[int, float]:
    """Add two numbers."""
    return a + b


def subtract(a: Union[int, float], b: Union[int, float]) -> Union[int, float]:
    """Subtract two numbers."""
    return a - b


def multiply(a: Union[int, float], b: Union[int, float]) -> Union[int, float]:
    """Multiply two numbers."""
    return a * b


def divide(a: Union[int, float], b: Union[int, float]) -> Union[int, float]:
    """Divide two numbers."""
    if b == 0:
        raise ValueError("Division by zero is not allowed")
    return a / b


def power(a: Union[int, float], b: Union[int, float]) -> Union[int, float]:
    """Raise a to the power of b."""
    return a ** b


def main():
    """Main entry point for the calculator tool."""
    try:
        # Read input from stdin
        input_data = sys.stdin.read()
        if not input_data.strip():
            raise ValueError("No input data provided")
        
        # Parse JSON input
        data = json.loads(input_data)
        params = data.get("params", {})
        context = data.get("context", {})
        
        # Extract operation and operands
        operation = params.get("operation")
        if not operation:
            raise ValueError("Operation parameter is required")
        
        a = params.get("a")
        b = params.get("b")
        
        if a is None or b is None:
            raise ValueError("Both 'a' and 'b' parameters are required")
        
        # Validate operands are numbers
        if not isinstance(a, (int, float)) or not isinstance(b, (int, float)):
            raise ValueError("Operands must be numbers")
        
        # Perform the operation
        operations = {
            "add": add,
            "subtract": subtract,
            "multiply": multiply,
            "divide": divide,
            "power": power,
        }
        
        if operation not in operations:
            raise ValueError(f"Unknown operation: {operation}. Supported operations: {list(operations.keys())}")
        
        result = operations[operation](a, b)
        
        # Prepare output
        output = {
            "result": result,
            "operation": operation,
            "operands": {"a": a, "b": b},
            "execution_id": context.get("execution_id"),
            "timestamp": context.get("started_at"),
        }
        
        # Write result to stdout as JSON
        print(json.dumps(output))
        
    except Exception as e:
        # Write error to stdout as JSON (not stderr, so the plugin can parse it)
        error_output = {
            "error": str(e),
            "error_type": type(e).__name__,
            "success": False,
        }
        print(json.dumps(error_output))
        sys.exit(1)


if __name__ == "__main__":
    main()

File: examples/nodejs_tools/simple_calculator.js
#!/usr/bin/env node

/**
 * Simple Calculator Tool for Node.js Plugin Example
 * 
 * This tool demonstrates basic arithmetic operations and shows how to:
 * - Read input from stdin as JSON
 * - Validate parameters
 * - Perform calculations
 * - Return results as JSON
 * - Handle errors gracefully
 */

const readline = require('readline');

// Create readline interface for reading from stdin
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Read all input from stdin
let inputData = '';
rl.on('line', (line) => {
    inputData += line + '\n';
});

rl.on('close', () => {
    try {
        // Parse input JSON
        const input = JSON.parse(inputData.trim());
        const params = input.params || {};
        const context = input.context || {};

        // Log context information for debugging
        if (context.execution_id) {
            console.error(`[DEBUG] Execution ID: ${context.execution_id}`);
        }

        // Validate required parameters
        if (!params.operation) {
            throw new Error('Missing required parameter: operation');
        }

        if (typeof params.a !== 'number' || typeof params.b !== 'number') {
            throw new Error('Parameters a and b must be numbers');
        }

        const { operation, a, b } = params;
        let result;

        // Perform calculation based on operation
        switch (operation.toLowerCase()) {
            case 'add':
            case '+':
                result = a + b;
                break;
            case 'subtract':
            case '-':
                result = a - b;
                break;
            case 'multiply':
            case '*':
                result = a * b;
                break;
            case 'divide':
            case '/':
                if (b === 0) {
                    throw new Error('Division by zero is not allowed');
                }
                result = a / b;
                break;
            case 'power':
            case '**':
                result = Math.pow(a, b);
                break;
            case 'modulo':
            case '%':
                if (b === 0) {
                    throw new Error('Modulo by zero is not allowed');
                }
                result = a % b;
                break;
            default:
                throw new Error(`Unsupported operation: ${operation}. Supported operations: add, subtract, multiply, divide, power, modulo`);
        }

        // Return success result
        const output = {
            success: true,
            result: result,
            operation: operation,
            operands: { a, b },
            timestamp: new Date().toISOString(),
            execution_id: context.execution_id || null
        };

        console.log(JSON.stringify(output, null, 2));
        process.exit(0);

    } catch (error) {
        // Return error result
        const errorOutput = {
            success: false,
            error: error.message,
            timestamp: new Date().toISOString(),
            execution_id: (inputData && JSON.parse(inputData).context?.execution_id) || null
        };

        console.log(JSON.stringify(errorOutput, null, 2));
        process.exit(1);
    }
});

// Handle process termination
process.on('SIGINT', () => {
    const errorOutput = {
        success: false,
        error: 'Process interrupted',
        timestamp: new Date().toISOString()
    };
    console.log(JSON.stringify(errorOutput, null, 2));
    process.exit(1);
});

process.on('SIGTERM', () => {
    const errorOutput = {
        success: false,
        error: 'Process terminated',
        timestamp: new Date().toISOString()
    };
    console.log(JSON.stringify(errorOutput, null, 2));
    process.exit(1);
});